nautobot/extras/models/customfields.py

```diff
@@ -325,8 +328,97 @@ def validate(self, value):
 
             # Validate selected choice
             if self.type == CustomFieldTypeChoices.TYPE_SELECT:
-                if value not in self.choices:
-                    raise ValidationError(f"Invalid choice ({value}). Available choices are: {', '.join(self.choices)}")
+                if value not in self.choices.values_list("value", flat=True):
+                    raise ValidationError(
+                        f"Invalid choice ({value}). Available choices are: {', '.join(self.choices.values_list('value', flat=True))}"
+                    )
+
+            if self.type == CustomFieldTypeChoices.TYPE_MULTISELECT:
+                if not set(value).issubset(self.choices.values_list("value", flat=True)):
+                    raise ValidationError(
+                        f"Invalid choice(s) ({value}). Available choices are: {', '.join(self.choices.values_list('value', flat=True))}"
+                    )
 
         elif self.required:
             raise ValidationError("Required field cannot be empty.")
+
+    def delete(self, *args, **kwargs):
+        """
+        Handle the cleanup of old custom field data when a CustomField is deleted.
+        """
+        content_types = set(self.content_types.values_list("pk", flat=True))
+
+        super().delete(*args, **kwargs)
+
+        delete_custom_field_data.delay(self.name, content_types)
```

Comment:
Just revisiting the discussion from yesterday: At first glance it seems a little weird to me to overload `delete()` to fire the background task vs. registering a `pre_delete` signal to call it? Since `delete()` isn't called during bulk operations (even tho we have limited use of those) and signals are, just curious about the design tradeoffs here.

The other concern that comes to mind is if there's a need for other internal apps to be aware of when custom field values are deleted. If so, signals might be the better path?

Reply:
So my opinion on signals is they should primarily only be used for cross-app decoupling. Whenever you use signals to implement logic that is entirely internal to a single app (as is the case here with deleting a custom field), I find that to be an anti-pattern because it does nothing more than obscure the code path.

> The other concern that comes to mind is if there's a need for other internal apps to be aware of when custom field values are deleted. If so, signals might be the better path?

Nothing stops another app from registering their own signal for this event on custom fields, and this would actually be the correct usage of a signal IMO.

The only exception to the above that I hold is for `m2m_changed` because it offers abilities that otherwise are hard to implement in other code paths. I think you basically said the same thing yesterday?

Bulk ORM operations is a valid concern, however, in nautobot we explicitly avoid any such usage primarily due to the changelogging implementation. Would love to discuss that, but I see that separate from this thread.

Reply:
Good stuff. Thanks!